File: hydroc/plot.c
#include <stdlib.h>
#include "solver.h"

#ifdef PGPLOT
#include <cpgplot.h>
#endif

#define NCONTOURS 5

int plot(float ***u, const int nx, const int ny) {

#ifdef PGPLOT
	const int ng=2;
	double xl, xr, yl, yr, dx, dy;
	float  denscontours[NCONTOURS], prescontours[NCONTOURS];
	float *dens, *pres;
	double maxd=-1e19, mind=+1e-19;
	double maxp=-1e19, minp=+1e-19;
	static int called = 0;
	int i, j, count;	
	float tr[6] = {0.,0.,0.,0.,0.,0.};
	
	xl = 0.; xr = nx-2*ng-1;
	yl = 0.; yr = ny-2*ng-1;
    dx = 1.; dy = 1.;

	dens = (float *)malloc((nx-2*ng)*(ny-2*ng)*sizeof(float));
	pres = (float *)malloc((nx-2*ng)*(ny-2*ng)*sizeof(float));
	count = 0;
	for (j=ny-ng; j>ng; j--) {
		for (i=ng; i<nx-ng; i++) {
			dens[count] = u[j][i][IDENS];
            float vx = u[j][i][IMOMX]/dens[count];
            float vy = u[j][i][IMOMY]/dens[count];
			pres[count] = (u[j][i][IENER] - 0.5*(vx*vx+vy*vy)*dens[count]);
			if (dens[count] > maxd) maxd = dens[count];
			if (dens[count] < mind) mind = dens[count];
			if (pres[count] > maxp) maxp = pres[count];
			if (pres[count] < minp) minp = pres[count];
			count++;
		}
	}
	
	tr[0] = xl; tr[3] = yl;
	tr[1] = dx; tr[5] = dy;

	for (i=0; i<NCONTOURS; i++) {
		denscontours[i] = mind + (i+1)*(maxd-mind)/(NCONTOURS+1);
		prescontours[i] = minp + (i+1)*(maxp-minp)/(NCONTOURS+1);
	}


	if (!called) {
		cpgbeg(0, "/XWINDOW", 1, 1);
		called = 1;
		cpgask(0);
	}
	cpgenv(xl, xr, yl, yr, 1, 1);

	cpgsci(2);
	cpgcont(dens, nx-2*ng, ny-2*ng, 1, nx-2*ng, 1, ny-2*ng, denscontours, NCONTOURS, tr);
	if (minp != maxp) {
		cpgsci(3);
	    cpgcont(pres, nx-2*ng, ny-2*ng, 1, nx-2*ng, 1, ny-2*ng, prescontours, NCONTOURS, tr);
	}
	cpgsci(1);
	
	free(dens); free(pres);
#endif
	return 0;
}

int closeplot(){
#ifdef PGPLOT
	cpgend();
#endif
	return 0;
}

File: hydrof/mpi/plot.f90
module plot

    integer, parameter, private :: ncontours=5
    public :: openplot, showplot, closeplot

contains
    subroutine openplot(nx, ny)
    use solver, only : nguard, idens, iener, imomx, imomy
    implicit none
    integer, intent(in) :: nx, ny
    real :: xl, xr, yl, yr

    xl = 1.
    yl = 1.
    xr = nx - 2*nguard 
    yr = ny - 2*nguard 

    call pgbeg(0, "/XWINDOW", 1, 1);
    call pgask(0)
    call pgenv(xl, xr, yl, yr, 1, 1)
    end subroutine openplot


    subroutine showplot(u, startn) 
    use solver
    implicit none
    real, dimension(:,:,:), intent(IN) :: u
    integer, intent(IN) :: startn
    integer :: nx, ny

    real, dimension(ncontours) :: denscontours, prescontours
    real, dimension(:,:), allocatable :: dens, pres
    real :: maxd, mind, maxp, minp
    real :: globmaxd, globmind, globmaxp, globminp
    real,dimension(6) :: tr
    integer :: i, ierr

    nx = size(u,2)
    ny = size(u,3)

    allocate(dens((nx-2*nguard),(ny-2*nguard)))
    allocate(pres((nx-2*nguard),(ny-2*nguard)))
    dens = u(idens, nguard+1:nx-nguard, nguard+1:ny-nguard)
    pres = u(iener, nguard+1:nx-nguard, nguard+1:ny-nguard) - &
              0.5*((u(imomx, nguard+1:nx-nguard, nguard+1:ny-nguard)**2)/dens + &
                   (u(imomy, nguard+1:nx-nguard, nguard+1:ny-nguard)**2)/dens)
    mind = minval(dens)
    maxd = maxval(dens)
    minp = minval(pres)
    maxp = maxval(pres)

    ! get globmind, globmaxd, mind, maxd from mind/maxd/minp/maxp
    globmind = mind
    globminp = minp
    globmaxd = maxd
    globmaxp = maxp

    tr = [startn*1.,1.,0.,0.,0.,1.]

    denscontours = [(globmind+(i*(globmaxd-globmind)/(ncontours+1)), i=1,ncontours)]
    prescontours = [(globminp+(i*(globmaxp-globminp)/(ncontours+1)), i=1,ncontours)]

    call pgenv(startn+1., startn+nx-2.*nguard , 1., ny-2.*nguard, 1, 1)
    call pgsci(2)
    call pgcont(dens, nx-2*nguard, ny-2*nguard, 1, nx-2*nguard, 1, ny-2*nguard, &
                denscontours, ncontours, tr)
    if (minp /= maxp) then
        call pgsci(3)
        call pgcont(pres, nx-2*nguard, ny-2*nguard, 1, nx-2*nguard, 1, ny-2*nguard, &
                    prescontours, ncontours, tr)
    endif
    call pgsci(1)
    deallocate(dens,pres)
    end subroutine showplot

    subroutine closeplot
    call pgend()
    end subroutine closeplot

end module plot
